connectors/sources/sharepoint_online.py

```diff
@@ -879,258 +995,283 @@ async def validate_config(self):
+        already_seen_usernames = set()
 
-        access_control = await self._access_control_for_role_assignments(
-            site_web_url, role_assignments
-        )
+        def _identity_mail(identity):
+            return identity.get("EMail", identity.get("mail", None))
 
-        return self._decorate_with_access_control(site_page, access_control)
+        def _identity_username(identity):
+            return identity.get("UserName", identity.get("userPrincipalName", None))
 
-    async def _access_control_for_role_assignments(
-        self, site_web_url, role_assignments
-    ):
-        access_control = []
+        def _already_seen(identity):
+            if _identity_mail(identity) in already_seen_emails:
+                return True
 
-        for role_assignment in role_assignments.get("value", []):
-            access_control += await self.client.users_and_groups_for_role_assignment(
-                site_web_url, role_assignment
-            )
+            if _identity_username(identity) in already_seen_usernames:
+                return True
 
-        return access_control
+            return False
 
-    async def _with_list_item_access_control(
-        self, site_web_url, site_list_name, list_item
-    ):
-        if not self._dls_enabled():
-            return list_item
+        def update_already_seen(identity):
+            mail = _identity_mail(identity)
+            username = _identity_username(identity)
 
-        list_item_role_assignments = await self.client.site_list_item_role_assignments(
-            site_web_url, site_list_name, list_item["id"]
-        )
+            # We want to make sure to not add 'None' to the already seen sets
+            if mail:
+                already_seen_emails.add(mail)
 
-        access_control = await self._access_control_for_role_assignments(
-            site_web_url, list_item_role_assignments
-        )
+            if username:
+                already_seen_usernames.add(username)
 
-        return self._decorate_with_access_control(list_item, access_control)
+        async for site_collection in self.client.site_collections():
+            async for site in self.client.sites(
+                site_collection["siteCollection"]["hostname"],
+                self.configuration["site_collections"],
+            ):
+                async for user_information in self.client.user_information_list(
+                    site["id"]
+                ):
+                    user = user_information["fields"]
 
-    def _dls_enabled(self):
-        if self._features is None:
-            return False
+                    if is_domain_group(user):
+                        domain_group_id = _domain_group_id(user["Name"])
 
-        if not self._features.document_level_security_enabled():
-            return False
+                        if domain_group_id:
+                            async for member_username in _emails_and_usernames_of_domain_group(
```

Comment:
Can `_emails_and_usernames_of_domain_group` yield a tuple of `email` and `username`, instead of one `email` and one `username`? So that we can check `_already_seen` without calling `self.client.user`?